### PR TITLE
Implementing FCM sendAll() API

### PR DESCRIPTION
**Pull request by hiranya911**
Implementing the following new API:

```
sendAll(messages: Message[], dryRun?: boolean): Promise<BatchResponse>
```

This implements a part of go/fcm-multicast. I've also implemented the following refactorings in order to keep the size of the `messaging.ts` module in check:

* Moved all the type definitions and their util functions to a new `messaging-types.ts` module.
* Moved the error handling/parsing logic to a new `messaging-errors.ts` module.

There's more room for refactoring here, but I would rather do that in separate PRs.

### CODE REVIEW

File: src/messaging/messaging-api-request.ts [hunk 1]
```diff
@@ -118,46 +82,51 @@ export class FirebaseMessagingRequestHandler {
...
-    if (err.response.isJson()) {
-      // For JSON responses, map the server response to a client-side error.
-      const json = err.response.data;
-      const errorCode = FirebaseMessagingRequestHandler.getErrorCode(json);
-      const errorMessage = FirebaseMessagingRequestHandler.getErrorMessage(json);
-      throw FirebaseMessagingError.fromServerError(errorCode, errorMessage, json);
-    }
+  /**
+   * Sends the given array of sub requests as a single batch to FCM, and parses the result into
+   * an BatchResponse object.
```
Comment (chong-shao) [inline src/messaging/messaging-api-request.ts]: s/an/a ?

File: src/messaging/messaging-errors.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+import {HttpError} from '../utils/api-request';
+import {FirebaseMessagingError, MessagingClientErrorCode} from '../utils/error';
+import * as validator from '../utils/validator';
+
+/**
+ * Creates a new FirebaseMessagingError by extracting the error code, message and other relevant
+ * details from an HTTP error response.
+ *
+ * @param {HttpError} err The HttpError to convert into a Firebase error
+ * @return {FirebaseMessagingError} A Fireabse error that can be returned to the user.
```
Comment (chong-shao) [inline src/messaging/messaging-errors.ts]: s/Fireabse/Firebase

File: src/messaging/messaging-api-request.ts (hunk 1 above)
Comment (hiranya911) [reply src/messaging/messaging-api-request.ts]: Done

File: src/messaging/messaging-errors.ts (hunk 2 above)
Comment (hiranya911) [reply src/messaging/messaging-errors.ts]: Done

File: src/messaging/messaging.ts [hunk 3]
```diff
@@ -811,6 +273,51 @@ export class Messaging implements FirebaseServiceInterface {
...
+   * @return {Promise<BatchResponse>} A Promise fulfilled with an object representing the result
+   *     of the send operation.
+   */
+  public sendAll(messages: Message[], dryRun?: boolean): Promise<BatchResponse> {
+    const copy: Message[] = deepCopy(messages);
+    if (!validator.isNonEmptyArray(copy)) {
+      throw new FirebaseMessagingError(
+        MessagingClientErrorCode.INVALID_ARGUMENT, 'messages must be a non-empty array');
+    }
+    if (copy.length > 100) {
```
Comment (bojeil-google) [inline src/messaging/messaging.ts]: Nit: Create a const for the maximum value.

File: src/utils/api-request.ts [hunk 4]
```diff
@@ -174,6 +207,57 @@ export class HttpClient {
...
+export function parseHttpResponse(
+  response: string | Buffer, config: HttpRequestConfig): HttpResponse {
+
+  const responseText: string = validator.isBuffer(response)
+    ? response.toString('utf-8') : response as string;
+  const endOfHeaderPos: number = responseText.indexOf('\r\n\r\n');
+  const headerLines: string[] = responseText.substring(0, endOfHeaderPos).split('\r\n');
+
+  const statusLine: string = headerLines[0];
+  const status: string = statusLine.trim().split(/\s/)[1];
```
Comment (bojeil-google) [inline src/utils/api-request.ts]: Validate that the array has >=2 elements and before that headerLines has at least 1 element.

File: src/utils/api-request.ts [hunk 5]
```diff
@@ -120,6 +122,37 @@ class DefaultHttpResponse implements HttpResponse {
...
+class MultipartHttpResponse implements HttpResponse {
+
+  public readonly status: number;
+  public readonly headers: any;
+  public readonly multipart: Buffer[];
+
+  constructor(resp: LowLevelResponse) {
+    this.status = resp.status;
+    this.headers = resp.headers;
+    this.multipart = resp.multipart;
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/utils/api-request.ts]: This class is not exported, and we control the only place where it gets instantiated. So it should be fine.

File: src/utils/api-request.ts [hunk 6]
```diff
@@ -264,22 +348,48 @@ function sendRequest(config: HttpRequestConfig): Promise<LowLevelResponse> {
       };
 
       const responseBuffer: Buffer[] = [];
-      respStream.on('data', (chunk: Buffer) => {
-        responseBuffer.push(chunk);
-      });
+      const boundary = getMultipartBoundary(res.headers);
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/utils/api-request.ts]: It's not easy due to the use of `new Promise()` but I think we can do some improvements. However, I'd rather do that as a separate PR.

File: src/utils/api-request.ts (hunk 4 above)
Comment (hiranya911) [reply src/utils/api-request.ts]: `slice(1)` returns an empty array when length is less than 2. So it's safe to iterate over it without a length check. Similarly, `headerLines` is the result of a `split()` call, which always returns an array with at least 1 element.

File: src/messaging/messaging.ts (hunk 3 above)
Comment (hiranya911) [reply src/messaging/messaging.ts]: Done

File: src/utils/api-request.ts [hunk 7]
```diff
@@ -289,18 +399,41 @@ function sendRequest(config: HttpRequestConfig): Promise<LowLevelResponse> {
...
   });
 }
 
+function getMultipartBoundary(headers: http.IncomingHttpHeaders): string {
+  const contentType = headers['content-type'];
+  if (!contentType.startsWith('multipart/')) {
+    return null;
+  }
+
+  const segments: string[] = contentType.split(';');
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/utils/api-request.ts]: Added the info to the jsdoc comment.

File: src/utils/api-request.ts [hunk 8]
```diff
@@ -174,6 +207,57 @@ export class HttpClient {
...
+ *
+ * @param {string|Buffer} response The HTTP response to be parsed.
+ * @param {HttpRequestConfig} config The request configuration that resulted in the HTTP response.
+ * @return {HttpResponse} An object containing the parsed HTTP status, headers and the body.
+ */
+export function parseHttpResponse(
+  response: string | Buffer, config: HttpRequestConfig): HttpResponse {
+
+  const responseText: string = validator.isBuffer(response)
+    ? response.toString('utf-8') : response as string;
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/utils/api-request.ts]: Done

File: src/utils/api-request.ts [hunk 9]
```diff
@@ -120,6 +122,37 @@ class DefaultHttpResponse implements HttpResponse {
   }
 }
 
+class MultipartHttpResponse implements HttpResponse {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/utils/api-request.ts]: Done

File: src/messaging/batch-request.ts [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+  part += `Content-Length: ${serializedRequest.length}\r\n`;
+  part += 'Content-Type: application/http\r\n';
+  part += `content-id: ${idx + 1}\r\n`;
+  part += 'content-transfer-encoding: binary\r\n';
+  part += '\r\n';
+  part += `${serializedRequest}\r\n`;
+  return part;
+}
+
+function serializeSubRequest(request: SubRequest): string {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/messaging/batch-request.ts]: Done

File: src/messaging/batch-request.ts [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+/**
+ * Represents a request that can be sent as part of an HTTP batch request.
+ */
+export interface SubRequest {
+  url: string;
+  body: object;
+  headers?: {[key: string]: any};
+}
+
+function createPart(request: SubRequest, delim: string, idx: number): string {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/messaging/batch-request.ts]: Done

File: src/messaging/batch-request.ts [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+/**
+ * Represents a request that can be sent as part of an HTTP batch request.
+ */
+export interface SubRequest {
+  url: string;
+  body: object;
+  headers?: {[key: string]: any};
+}
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/messaging/batch-request.ts]: Done

**Comment by hiranya911** [comment]
Thanks @bojeil-google. I've responded to your comments. PTAL.

File: test/unit/messaging/batch-requests.spec.ts [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+    const args: HttpRequestConfig = stub.getCall(0).args[0];
+    expect(args.headers).to.have.property('X-Custom-Header', 'value');
+  });
+
+  it('should add sub request headers to the payload', async () => {
+    const stub = sinon.stub(httpClient, 'send').resolves(
+      createMultipartResponse([responseObject]));
+    stubs.push(stub);
+    const requests: SubRequest[] = [
+      {url: 'https://example.com', body: {foo: 1}, headers: {'X-Custom-Header': 'value'}},
```
Comment (bojeil-google) [inline test/unit/messaging/batch-requests.spec.ts]: The test uses the same header key/value for the common headers and the subrequests. it is better to use different values to confirm source of expected value.

File: test/unit/messaging/batch-requests.spec.ts [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+    const args: HttpRequestConfig = stub.getCall(0).args[0];
+    const parsedRequest = parseHttpRequest(args.data as Buffer);
+    expect(parsedRequest.multipart.length).to.equal(requests.length);
+    parsedRequest.multipart.forEach((part: {body: Buffer}) => {
+      const parsedPart: {headers: object} = parseHttpRequest(part.body.toString().trim());
+      expect(parsedPart.headers).to.have.property('X-Custom-Header', 'value');
+    });
+  });
+
+  function checkOutgoingRequest(stub: sinon.SinonStub, requests: SubRequest[]) {
```
Comment (bojeil-google) [inline test/unit/messaging/batch-requests.spec.ts]: nit: Add helper functions to the top of the file. It is easier to read sequentially.

File: test/unit/messaging/batch-requests.spec.ts [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+      requests.forEach((req, idx) => {
+        const part = parsedRequest.multipart[idx].body.toString().trim();
+        expect(part).to.equal(getParsedPartData(req.body));
+      });
+    }
+  }
+});
+
+function parseHttpRequest(text: string | Buffer): any {
+  const httpMessageParser = require('http-message-parser');
```
Comment (bojeil-google) [inline test/unit/messaging/batch-requests.spec.ts]: Suggestion: I think we should minimize depending on low usage npm modules. In this case, this has ~2k weekly downloads. We have been bitten by some dependency on an npm module that was abandoned in the past.

File: test/unit/messaging/messaging.spec.ts [hunk 16]
```diff
@@ -463,6 +530,230 @@ describe('Messaging', () => {
...
+      ).should.eventually.be.rejected.and.have.property('code', 'messaging/invalid-argument');
+    });
+
+    it('should be rejected given an app which returns null access tokens', () => {
+      return nullAccessTokenMessaging.sendAll(
+        [validMessage],
+      ).should.eventually.be.rejected.and.have.property('code', 'app/invalid-credential');
+    });
+
+    function checkSendResponseSuccess(response: SendResponse, messageId: string) {
```
Comment (bojeil-google) [inline test/unit/messaging/messaging.spec.ts]: nit: define helper functions on top of file.

File: test/unit/utils/api-request.spec.ts [hunk 17]
```diff
@@ -430,6 +583,51 @@ describe('HttpClient', () => {
       expect(resp.data).to.deep.equal(respData);
     });
   });
+
+  it('should reject if the request payload is invalid', () => {
+    const client = new HttpClient();
+    const err = 'Error while making request: Request data must be a string, a Buffer '
+     + 'or a json serializable object';
```
Comment (bojeil-google) [inline test/unit/utils/api-request.spec.ts]: fix indentation

File: test/unit/utils/api-request.spec.ts [hunk 18]
```diff
@@ -681,3 +879,93 @@ describe('ApiSettings', () => {
...
+    expect(response.status).to.equal(200);
+    expect(Object.keys(response.headers).length).to.equal(2);
+    expect(response.headers).to.have.property('content-type', 'application/json');
+    expect(response.headers).to.have.property('date', 'Thu, 07 Feb 2019 19:20:34 GMT');
+    expect(response.isJson()).to.be.true;
+    expect(response.data).to.deep.equal({foo: 1});
+    expect(response.text).to.equal('{"foo": 1}');
+  });
+
+  it('should parse a error response with json content', () => {
```
Comment (bojeil-google) [inline test/unit/utils/api-request.spec.ts]: an error

File: test/unit/messaging/batch-requests.spec.ts (hunk 15 above)
Comment (hiranya911) [reply test/unit/messaging/batch-requests.spec.ts]: We only use this in a couple of test cases. I can try to replace it with something else in a future PR.

File: src/messaging/batch-request.ts [hunk 19]
```diff
@@ -70,25 +79,26 @@ export class BatchRequestClient {
...
 /**
- * Represents a request that can be sent as part of an HTTP batch request.
+ * Creates a single part in a multipart HTTP request body. The part consists of several headers
+ * followed by the serialized sub request as the body. As per the requirements of the FCM batch
+ * API, sets the content-type header to application/http, and the content-transfer-encoding to
+ * binary.
+ *
+ * @param {SubRequest} request A sub request that will be used to populate the part.
+ * @param {string} boundary Multipart boundary string.
+ * @param {number} idx An index number that is used to set the content-id header.
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/messaging/batch-request.ts]: Done

File: src/messaging/batch-request.ts [hunk 20]
```diff
@@ -98,6 +108,13 @@ function createPart(request: SubRequest, delim: string, idx: number): string {
   return part;
 }
 
+/**
+ * Serializes a sub request into a string that can be embedded in a multipart HTTP request. The
+ * format of the string is the wire format of a typical HTTP request, consisting of a header and a
+ * body.
+ *
+ * @param request {SubRequest} The sub request to be serialized.
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/messaging/batch-request.ts]: Done

File: src/utils/api-request.ts [hunk 21]
```diff
@@ -289,18 +403,46 @@ function sendRequest(config: HttpRequestConfig): Promise<LowLevelResponse> {
...
     // Send the request
     req.end(data);
   });
 }
 
+/**
+ * Extracts multipart boundary from the HTTP header. The content-type header of a multipart
+ * response has the form 'multipart/subtype; boundary=string'.
+ */
+function getMultipartBoundary(headers: http.IncomingHttpHeaders): string {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/utils/api-request.ts]: Done (although it seems like TS compiler doesn't really enforce this by default)

File: test/unit/messaging/batch-requests.spec.ts (hunk 13 above)
Comment (hiranya911) [reply test/unit/messaging/batch-requests.spec.ts]: Done

File: test/unit/messaging/batch-requests.spec.ts (hunk 14 above)
Comment (hiranya911) [reply test/unit/messaging/batch-requests.spec.ts]: Done

File: test/unit/messaging/messaging.spec.ts (hunk 16 above)
Comment (hiranya911) [reply test/unit/messaging/messaging.spec.ts]: Done. Moved to the top of the enclosing block. These helpers are scoped to a specific test case.

File: test/unit/utils/api-request.spec.ts (hunk 17 above)
Comment (hiranya911) [reply test/unit/utils/api-request.spec.ts]: Done

File: test/unit/utils/api-request.spec.ts (hunk 18 above)
Comment (hiranya911) [reply test/unit/utils/api-request.spec.ts]: Done

File: src/utils/api-request.ts (hunk 4 above)
Comment (hiranya911) [reply src/utils/api-request.ts]: We use split and slice for `headerLines`. For `statusLine` I added a check to cover the case of a malformed header. Although chances of it happening should be negligibly small.

**Comment by hiranya911** [comment]
Thanks @bojeil-google. I've addressed most of your comments.

**Pull request merged**